Adds copyStream buffer pool

AlexanderYastrebov opened the pull request:
Reduces memory allocation by reusing byte slices in `copyStream`.

**Profiling results**

```
bin/skipper -inline-routes='* -> inlineContent("Hello") -> <shunt>' -access-log-disabled -enable-profile
```

Baseline (`master`)
```
$ echo "GET http://localhost:9090/test" | vegeta attack -duration=10m -rate=0 --max-workers=500 | vegeta report
Requests      [total, rate, throughput]         16754506, 27924.03, 27923.66
Duration      [total, attack, wait]             10m0s, 10m0s, 8.097ms
Latencies     [min, mean, 50, 90, 95, 99, max]  89.763µs, 14.278ms, 9.62ms, 36.179ms, 43.932ms, 65.012ms, 228.741ms
Bytes In      [total, mean]                     83772530, 5.00
Bytes Out     [total, mean]                     0, 0.00
Success       [ratio]                           100.00%
Status Codes  [code:count]                      200:16754506
Error Set:
```
```
$ go tool pprof -top http://localhost:9911/debug/pprof/allocs
...
Showing nodes accounting for 194.44GB, 96.27% of 201.96GB total
Dropped 137 nodes (cum <= 1.01GB)
      flat  flat%   sum%        cum   cum%
  127.97GB 63.36% 63.36%      128GB 63.38%  github.com/zalando/skipper/proxy.copyStream
    9.04GB  4.48% 67.84%    10.13GB  5.01%  github.com/zalando/skipper/filters/builtin.(*inlineContent).Request
    8.16GB  4.04% 71.88%     8.43GB  4.17%  net/textproto.(*Reader).ReadMIMEHeader
       7GB  3.47% 75.35%        7GB  3.47%  net/http.Header.Clone
    5.95GB  2.95% 78.30%    23.89GB 11.83%  net/http.(*conn).readRequest
    5.46GB  2.70% 81.00%     5.46GB  2.70%  github.com/zalando/skipper/proxy.copyHeader
       5GB  2.47% 83.48%     5.12GB  2.53%  github.com/zalando/skipper/proxy.newContext
    4.02GB  1.99% 85.47%     6.21GB  3.08%  net/http.(*Request).WithContext (inline)
    4.01GB  1.99% 87.45%    15.52GB  7.69%  net/http.readRequest
    3.96GB  1.96% 89.41%     3.96GB  1.96%  github.com/zalando/skipper/proxy.(*proxyTracing).logEvent (inline)
    2.48GB  1.23% 90.64%     2.54GB  1.26%  fmt.Sprintf
    2.27GB  1.12% 91.76%     2.27GB  1.12%  net/url.parse
    2.19GB  1.09% 92.85%     2.19GB  1.09%  net/http.cloneURL (inline)
    1.46GB  0.72% 93.57%     1.46GB  0.72%  github.com/zalando/skipper/tracing.CreateSpan
    1.27GB  0.63% 94.20%     1.27GB  0.63%  context.WithCancel
    1.26GB  0.63% 94.82%     1.26GB  0.63%  net.(*conn).Read
    1.19GB  0.59% 95.41%     1.38GB  0.68%  github.com/zalando/skipper/proxy.(*Proxy).setCommonSpanInfo
    1.14GB  0.56% 95.98%     1.14GB  0.56%  bufio.NewWriterSize (inline)
    0.47GB  0.23% 96.21%   176.60GB 87.44%  github.com/zalando/skipper/proxy.(*Proxy).ServeHTTP
    0.13GB 0.064% 96.27%    14.74GB  7.30%  github.com/zalando/skipper/proxy.(*Proxy).applyFiltersToRequest
...
```

This change:
```
$ echo "GET http://localhost:9090/test" | vegeta attack -duration=10m -rate=0 --max-workers=500 | vegeta report
Requests      [total, rate, throughput]         18090343, 30149.97, 30149.93
Duration      [total, attack, wait]             10m0s, 10m0s, 867.153µs
Latencies     [min, mean, 50, 90, 95, 99, max]  74.27µs, 10.584ms, 8.151ms, 23.656ms, 29.943ms, 42.609ms, 132.707ms
Bytes In      [total, mean]                     90451715, 5.00
Bytes Out     [total, mean]                     0, 0.00
Success       [ratio]                           100.00%
Status Codes  [code:count]                      200:18090343
Error Set:
```
```
$ go tool pprof -top http://localhost:9911/debug/pprof/allocs
...
Showing nodes accounting for 76.82GB, 97.85% of 78.51GB total
Dropped 100 nodes (cum <= 0.39GB)
      flat  flat%   sum%        cum   cum%
    9.74GB 12.41% 12.41%    10.85GB 13.82%  github.com/zalando/skipper/filters/builtin.(*inlineContent).Request
    8.84GB 11.26% 23.67%     9.10GB 11.59%  net/textproto.(*Reader).ReadMIMEHeader
    7.47GB  9.52% 33.18%     7.47GB  9.52%  net/http.Header.Clone
    6.45GB  8.22% 41.40%    24.42GB 31.10%  net/http.(*conn).readRequest
    5.94GB  7.56% 48.96%     5.94GB  7.56%  github.com/zalando/skipper/proxy.copyHeader
    5.35GB  6.82% 55.78%     5.49GB  6.99%  github.com/zalando/skipper/proxy.newContext
    4.32GB  5.50% 61.28%     6.76GB  8.61%  net/http.(*Request).WithContext (inline)
    4.31GB  5.49% 66.77%     4.31GB  5.49%  github.com/zalando/skipper/proxy.(*proxyTracing).logEvent (inline)
    4.29GB  5.46% 72.23%    16.53GB 21.06%  net/http.readRequest
    2.68GB  3.41% 75.64%     2.70GB  3.44%  fmt.Sprintf
    2.46GB  3.14% 78.78%     2.46GB  3.14%  net/url.parse
    2.45GB  3.11% 81.89%     2.45GB  3.11%  net/http.cloneURL (inline)
    1.63GB  2.07% 83.96%     1.63GB  2.07%  github.com/zalando/skipper/tracing.CreateSpan
    1.36GB  1.73% 85.69%     1.36GB  1.73%  net.(*conn).Read
    1.33GB  1.70% 87.39%     1.34GB  1.70%  context.WithCancel
    1.32GB  1.69% 89.08%     1.52GB  1.94%  github.com/zalando/skipper/proxy.(*Proxy).setCommonSpanInfo
    1.09GB  1.39% 90.47%     1.09GB  1.39%  github.com/zalando/skipper/proxy.(*context).setMetricsPrefix (inline)
    0.86GB  1.09% 91.56%     0.86GB  1.09%  bytes.NewBufferString (inline)
    0.83GB  1.06% 92.61%     0.83GB  1.06%  github.com/zalando/skipper/proxy.appendParams
    0.81GB  1.03% 93.65%     0.81GB  1.03%  github.com/zalando/skipper/routing.(*matcher).match
    0.76GB  0.97% 94.62%     0.76GB  0.97%  context.WithValue
    0.55GB   0.7% 95.32%     0.55GB   0.7%  net/textproto.(*Reader).ReadLine (inline)
    0.54GB  0.69% 96.01%    52.58GB 66.97%  github.com/zalando/skipper/proxy.(*Proxy).ServeHTTP
    0.52GB  0.66% 96.67%     0.52GB  0.66%  github.com/zalando/skipper/logging.NewLoggingWriter (inline)
    0.27GB  0.35% 97.02%     0.80GB  1.01%  github.com/zalando/skipper/metrics.(*CodaHale).MeasureFilterResponse
    0.26GB  0.33% 97.35%     0.81GB  1.03%  github.com/zalando/skipper/metrics.(*CodaHale).MeasureResponse
    0.25GB  0.32% 97.67%     0.78GB     1%  github.com/zalando/skipper/metrics.(*CodaHale).MeasureFilterRequest
    0.14GB  0.18% 97.85%    15.82GB 20.15%  github.com/zalando/skipper/proxy.(*Proxy).applyFiltersToRequest
...
```

szuecs commented on proxy/proxy.go:
```diff
@@ -406,10 +407,29 @@ func cloneHeaderExcluding(h http.Header, excludeList map[string]bool) http.Heade
+//
+// Note: sync.Pool example suggests to use bytes.Buffer that can dynamically grow and thus
+// cause problems, see https://github.com/golang/go/issues/23199 for details.
+var bufferPool = sync.Pool{
```
Can you change it to non package global?
It’s a bad thing that we used it and we should omit it if we can.

AlexanderYastrebov replied:
It is possible to make an instance per proxy. What is the problem with single shared pool (even for multiple proxies)?

AlexanderYastrebov replied:
Another thing is that maybe it should be under the flag to turn it on/off quickly to experiment of if there is some issue with it in real setup...

szuecs replied:
Everytime we had globals it turned out to be bad. We should avoid creating them.
Of course we have globals that we should some day get rid of.

AlexanderYastrebov replied:
Will fix it, no problem, was just genuinely interested what could be the possible problem with it.

aryszka replied:
One typical regret is when testing a previously unanticipated scenario.

Though true, that a bufferpool is kind of just a reference to the process's actual memory pool, and no actual problem comes to my mind right now, but still, I also prefer to just avoid globals, especially as it's easy enough.

AlexanderYastrebov commented:
I would like to run additional profile session with increasing payload sizes.

AlexanderYastrebov commented:
I did some more tests with 4 response sizes served as files:
```
$ truncate -s 1k 1k.bin
$ truncate -s 8k 8k.bin
$ truncate -s 16k 16k.bin
$ truncate -s 32k 32k.bin
$ bin/skipper -inline-routes='* -> static("/", "./") -> <shunt>' -access-log-disabled -enable-profile
```
Overall less memory is allocated with buffer as expected, latencies seem to improve a bit as well.
`static` filter is using `http.FileServer` which (`io.copyBuffer`) starts to dominate in allocation for bigger filesizes.

<details>

```
$ # master 1k
$ echo "GET http://localhost:9090/1k.bin" | vegeta attack -duration=10m -rate=0 --max-workers=500 | vegeta report && go tool pprof -top http://localhost:9911/debug/pprof/allocs
Requests      [total, rate, throughput]         11891224, 19818.63, 19818.35
Duration      [total, attack, wait]             10m0s, 10m0s, 8.549ms
Latencies     [min, mean, 50, 90, 95, 99, max]  128.034µs, 19.777ms, 12.265ms, 48.826ms, 66.23ms, 99.468ms, 394.601ms
Bytes In      [total, mean]                     12176613376, 1024.00
Bytes Out     [total, mean]                     0, 0.00
Success       [ratio]                           100.00%
Status Codes  [code:count]                      200:11891224  
Error Set:
...
Showing nodes accounting for 164.80GB, 96.80% of 170.24GB total
Dropped 113 nodes (cum <= 0.85GB)
      flat  flat%   sum%        cum   cum%
   90.77GB 53.32% 53.32%    90.79GB 53.33%  github.com/zalando/skipper/proxy.copyStream
   11.33GB  6.66% 59.98%    11.38GB  6.69%  io.copyBuffer
    5.76GB  3.38% 63.36%     5.94GB  3.49%  net/textproto.(*Reader).ReadMIMEHeader
...
```

```
$ # copy-stream-buffer-pool 1k
$ echo "GET http://localhost:9090/1k.bin" | vegeta attack -duration=10m -rate=0 --max-workers=500 | vegeta report && go tool pprof -top http://localhost:9911/debug/pprof/allocs
Requests      [total, rate, throughput]         13104733, 21841.23, 21840.48
Duration      [total, attack, wait]             10m0s, 10m0s, 20.461ms
Latencies     [min, mean, 50, 90, 95, 99, max]  120.95µs, 14.16ms, 10.192ms, 32.829ms, 40.881ms, 61.863ms, 323.921ms
Bytes In      [total, mean]                     13419246592, 1024.00
Bytes Out     [total, mean]                     0, 0.00
Success       [ratio]                           100.00%
Status Codes  [code:count]                      200:13104733  
Error Set:
...
Showing nodes accounting for 84644.57MB, 95.37% of 88754.09MB total
Dropped 109 nodes (cum <= 443.77MB)
      flat  flat%   sum%        cum   cum%
12825.52MB 14.45% 14.45% 12863.52MB 14.49%  io.copyBuffer
 6616.06MB  7.45% 21.90%  6822.57MB  7.69%  net/textproto.(*Reader).ReadMIMEHeader
 6015.09MB  6.78% 28.68%  6015.09MB  6.78%  net/http.Header.Clone
...
```

```
$ # master 8k
$ echo "GET http://localhost:9090/8k.bin" | vegeta attack -duration=10m -rate=0 --max-workers=500 | vegeta report && go tool pprof -top http://localhost:9911/debug/pprof/allocs
Requests      [total, rate, throughput]         8747273, 14578.78, 14578.71
Duration      [total, attack, wait]             10m0s, 10m0s, 3.101ms
Latencies     [min, mean, 50, 90, 95, 99, max]  135.901µs, 15.303ms, 11.137ms, 36.137ms, 44.836ms, 63.761ms, 182.117ms
Bytes In      [total, mean]                     71657660416, 8192.00
Bytes Out     [total, mean]                     0, 0.00
Success       [ratio]                           100.00%
Status Codes  [code:count]                      200:8747273  
Error Set:
...
Showing nodes accounting for 175.61GB, 95.57% of 183.74GB total
Dropped 115 nodes (cum <= 0.92GB)
      flat  flat%   sum%        cum   cum%
   67.35GB 36.66% 36.66%    67.37GB 36.67%  github.com/zalando/skipper/proxy.copyStream
   66.77GB 36.34% 73.00%    66.78GB 36.35%  io.copyBuffer
    4.41GB  2.40% 75.40%     4.54GB  2.47%  net/textproto.(*Reader).ReadMIMEHeader
...
```

```
$ # copy-stream-buffer-pool 8k
$ echo "GET http://localhost:9090/8k.bin" | vegeta attack -duration=10m -rate=0 --max-workers=500 | vegeta report && go tool pprof -top http://localhost:9911/debug/pprof/allocs
Requests      [total, rate, throughput]         9011236, 15018.48, 15018.19
Duration      [total, attack, wait]             10m0s, 10m0s, 11.637ms
Latencies     [min, mean, 50, 90, 95, 99, max]  133.575µs, 12.839ms, 8.53ms, 31.477ms, 38.178ms, 52.27ms, 196.311ms
Bytes In      [total, mean]                     73820045312, 8192.00
Bytes Out     [total, mean]                     0, 0.00
Success       [ratio]                           100.00%
Status Codes  [code:count]                      200:9011236  
Error Set:
...
Showing nodes accounting for 118280.05MB, 95.88% of 123366.35MB total
Dropped 128 nodes (cum <= 616.83MB)
      flat  flat%   sum%        cum   cum%
70501.94MB 57.15% 57.15% 70520.44MB 57.16%  io.copyBuffer
 4546.59MB  3.69% 60.83%  4679.09MB  3.79%  net/textproto.(*Reader).ReadMIMEHeader
 4123.09MB  3.34% 64.18%  4123.09MB  3.34%  net/http.Header.Clone
...
```

```
$ # master 16k
$ echo "GET http://localhost:9090/16k.bin" | vegeta attack -duration=10m -rate=0 --max-workers=500 | vegeta report && go tool pprof -top http://localhost:9911/debug/pprof/allocs
Requests      [total, rate, throughput]         7504819, 12508.00, 12507.99
Duration      [total, attack, wait]             10m0s, 10m0s, 448.849µs
Latencies     [min, mean, 50, 90, 95, 99, max]  138.415µs, 16.006ms, 11.477ms, 38.454ms, 47.977ms, 67.097ms, 191.249ms
Bytes In      [total, mean]                     122958954496, 16384.00
Bytes Out     [total, mean]                     0, 0.00
Success       [ratio]                           100.00%
Status Codes  [code:count]                      200:7504819  
Error Set:
...
Showing nodes accounting for 210440.53MB, 95.69% of 219911.93MB total
Dropped 119 nodes (cum <= 1099.56MB)
      flat  flat%   sum%        cum   cum%
117194.21MB 53.29% 53.29% 117208.71MB 53.30%  io.copyBuffer
58929.12MB 26.80% 80.09% 58962.14MB 26.81%  github.com/zalando/skipper/proxy.copyStream
 3724.40MB  1.69% 81.78%  3845.90MB  1.75%  net/textproto.(*Reader).ReadMIMEHeader
...
```

```
$ # copy-stream-buffer-pool 16k
$ echo "GET http://localhost:9090/16k.bin" | vegeta attack -duration=10m -rate=0 --max-workers=500 | vegeta report && go tool pprof -top http://localhost:9911/debug/pprof/allocs
Requests      [total, rate, throughput]         6511975, 10853.28, 10853.27
Duration      [total, attack, wait]             10m0s, 10m0s, 642.461µs
Latencies     [min, mean, 50, 90, 95, 99, max]  148.599µs, 15.718ms, 9.62ms, 39.599ms, 51.205ms, 75.48ms, 230.647ms
Bytes In      [total, mean]                     106692198400, 16384.00
Bytes Out     [total, mean]                     0, 0.00
Success       [ratio]                           100.00%
Status Codes  [code:count]                      200:6511975  
Error Set:
...
Showing nodes accounting for 133307.87MB, 95.11% of 140168.22MB total
Dropped 116 nodes (cum <= 700.84MB)
      flat  flat%   sum%        cum   cum%
101809.33MB 72.63% 72.63% 101822.33MB 72.64%  io.copyBuffer
 3220.26MB  2.30% 74.93%  3324.77MB  2.37%  net/textproto.(*Reader).ReadMIMEHeader
 2985.79MB  2.13% 77.06%  2985.79MB  2.13%  net/http.Header.Clone
...
```

```
$ # master 32k
$ echo "GET http://localhost:9090/32k.bin" | vegeta attack -duration=10m -rate=0 --max-workers=500 | vegeta report && go tool pprof -top http://localhost:9911/debug/pprof/allocs
Requests      [total, rate, throughput]         5817260, 9695.43, 9695.16
Duration      [total, attack, wait]             10m0s, 10m0s, 17.193ms
Latencies     [min, mean, 50, 90, 95, 99, max]  190.429µs, 15.841ms, 9.603ms, 39.658ms, 52.541ms, 79.688ms, 226.15ms
Bytes In      [total, mean]                     190619975680, 32768.00
Bytes Out     [total, mean]                     0, 0.00
Success       [ratio]                           100.00%
Status Codes  [code:count]                      200:5817260  
Error Set:
...
Showing nodes accounting for 251863.23MB, 96.24% of 261695.81MB total
Dropped 122 nodes (cum <= 1308.48MB)
      flat  flat%   sum%        cum   cum%
181639.81MB 69.41% 69.41% 181648.81MB 69.41%  io.copyBuffer
45679.98MB 17.46% 86.86% 45731.99MB 17.48%  github.com/zalando/skipper/proxy.copyStream
 2919.70MB  1.12% 87.98%  3004.20MB  1.15%  net/textproto.(*Reader).ReadMIMEHeader
...
```

```
$ # copy-stream-buffer-pool 32k
$ echo "GET http://localhost:9090/32k.bin" | vegeta attack -duration=10m -rate=0 --max-workers=500 | vegeta report && go tool pprof -top http://localhost:9911/debug/pprof/allocs
Requests      [total, rate, throughput]         5837978, 9729.86, 9729.82
Duration      [total, attack, wait]             10m0s, 10m0s, 2.552ms
Latencies     [min, mean, 50, 90, 95, 99, max]  191.136µs, 13.612ms, 7.824ms, 34.232ms, 46.246ms, 72.586ms, 242.686ms
Bytes In      [total, mean]                     191298863104, 32768.00
Bytes Out     [total, mean]                     0, 0.00
Success       [ratio]                           100.00%
Status Codes  [code:count]                      200:5837978  
Error Set:
...
Showing nodes accounting for 208616.59MB, 95.72% of 217936.67MB total
Dropped 125 nodes (cum <= 1089.68MB)
      flat  flat%   sum%        cum   cum%
182327.87MB 83.66% 83.66% 182337.38MB 83.67%  io.copyBuffer
 2889.69MB  1.33% 84.99%  2974.19MB  1.36%  net/textproto.(*Reader).ReadMIMEHeader
 2680.71MB  1.23% 86.22%  2680.71MB  1.23%  net/http.Header.Clone
...
```

</details>

aryszka commented on proxy/proxy.go:
```diff
@@ -406,10 +407,29 @@ func cloneHeaderExcluding(h http.Header, excludeList map[string]bool) http.Heade
+// cause problems, see https://github.com/golang/go/issues/23199 for details.
+var bufferPool = sync.Pool{
+	New: func() interface{} {
+		b := make([]byte, proxyBufferSize)
```
this can be good point to limit buffer allocation in a later pass

aryszka commented:
apart of the global var, lgtm

AlexanderYastrebov commented:
Closing for now, needs benchmark reevaluation. The work was not a waste as it resulted in #1570 and #1616 improvements.
